Add smart pointer expectations to WebKit and WebCore

briannafan opened the pull request:
#### 7c786c8ebb152ce2980198a3c9c183b27e169a66
<pre>
Add smart pointer expectations to WebKit and WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=274320">https://bugs.webkit.org/show_bug.cgi?id=274320</a>
<a href="https://rdar.apple.com/128281857">rdar://128281857</a>

Reviewed by David Kilzer.

Revises scripts to save unexpected lists to smart pointer result archive.
Adds a section to the results page that links to the unexpected lists.
Creates DisplayUnexpectedResults step to show unexpected lists in Buildbot UI.

* Tools/Scripts/compare-static-analysis-results:
(find_diff):
(compare_project_results_to_expectations): Sorts unexpected lists alphabetically and saves in archive.
(compare_project_results_by_run): Removes .txt extension.

* Tools/Scripts/generate-dirty-files:
(find_project_results): Sorts filenames alphabetically. Removes .txt extension.

* Tools/Scripts/generate-static-analysis-archive:
(parse_command_line):
(generate_unexpected_results_links): Added.
(generate_results_page): If there are unexpected results, adds links to unexpected lists.

* Tools/CISupport/build-webkit-org/steps.py:
(AddToLogMixin): Added.
(CompareStaticAnalyzerResults): Creates unexpected_results.json.
(DisplayUnexpectedResults): Reads unexpected_results.json and creates logs.
(ExtractStaticAnalyzerTestResults): Invokes DisplayUnexpectedResults conditionally.

* Source/WebCore/SmartPointerExpectations: Added.
* Source/WebKit/SmartPointerExpectations: Added.
- Adds expectations for each checker type.

Canonical link: <a href="https://commits.webkit.org/279827@main">https://commits.webkit.org/279827@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e5efb240b8fa351ad97e06af1019ee8cdca33ed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54676 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34099 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7252 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57956 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/5409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56977 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41645 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5418 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/44289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/5409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56771 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/32239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/47359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/25414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/29035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/4694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3549 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/4906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59546 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/29917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/5056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/51711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/54674 "Passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/31062 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/47444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/32051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8085 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/30847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebKit/SmartPointerExpectations/RefCntblBaseVirtualDtorExpectations:
```diff
@@ -0,0 +1 @@
+GeneratedSerializers.mm
```
I wonder if we should name the `SmartPointerExpectations/` folder as `ClangStaticAnalyzerExpectations/SmartPointer/` since we will eventually want to add more checkers later.

However, we can update this in the future—no need to get this "perfect" now.

briannafan replied:
Noted, thanks!

ddkilzer commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -1780,6 +1788,40 @@ def getResultSummary(self):
+        yield self.getFilesPerProject(unexpected_results_data, 'failures')
+        if self.getProperty('unexpected_failing_files', 0) or self.getProperty('unexpected_passing_files', 0):
+            defer.returnValue(FAILURE)
+        defer.returnValue(SUCCESS)
```
I don't know Buildbot Python very well, but do you need a `return` statement here so that this code doesn't call `defer.returnValue(SUCCESS)` every time?

```suggestion
        if self.getProperty('unexpected_failing_files', 0) or self.getProperty('unexpected_passing_files', 0):
            defer.returnValue(FAILURE)
            return
        defer.returnValue(SUCCESS)
```

ddkilzer commented on Tools/Scripts/generate-static-analysis-archive:
```diff
@@ -57,17 +58,47 @@ def parse_command_line(args):
-                project_issue_count=get_project_issue_count_as_string(static_analyzer_index_path),
-                project_name=project_name)
+        project_count = get_project_issue_count_as_string(static_analyzer_index_path)
+        if project_count != '0':
```
Would it be better to convert `project_count` to a numeric value and then compare to `0` instead of the string `'0'`?

(If comparing to string `'0'` is done elsewhere, then you can ignore this feedback.)

ddkilzer reviewed (approved):
r=me.  This seems fine to me overall.  Just had a few minor questions noted below.

At this stage of the project, I think it's more important to get something working and then to iterate on it.

briannafan committed:
Add smart pointer expectations to WebKit and WebCore
https://bugs.webkit.org/show_bug.cgi?id=274320
rdar://128281857

Reviewed by David Kilzer.

Revises scripts to save unexpected lists to smart pointer result archive.
Adds a section to the results page that links to the unexpected lists.
Creates DisplayUnexpectedResults step to show unexpected lists in Buildbot UI.

* Tools/Scripts/compare-static-analysis-results:
(find_diff):
(compare_project_results_to_expectations): Sorts unexpected lists alphabetically and saves in archive.
(compare_project_results_by_run): Removes .txt extension.

* Tools/Scripts/generate-dirty-files:
(find_project_results): Sorts filenames alphabetically. Removes .txt extension.

* Tools/Scripts/generate-static-analysis-archive:
(parse_command_line):
(generate_unexpected_results_links): Added.
(generate_results_page): If there are unexpected results, adds links to unexpected lists.

* Tools/CISupport/build-webkit-org/steps.py:
(AddToLogMixin): Added.
(CompareStaticAnalyzerResults): Creates unexpected_results.json.
(DisplayUnexpectedResults): Reads unexpected_results.json and creates logs.
(ExtractStaticAnalyzerTestResults): Invokes DisplayUnexpectedResults conditionally.

* Source/WebCore/SmartPointerExpectations: Added.
* Source/WebKit/SmartPointerExpectations: Added.
- Adds expectations for each checker type.

Canonical link: https://commits.webkit.org/279827@main
